### PR TITLE
Fix for generating vertex data more efficiently

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: defold-spine/commonsrc/vertices.cpp [hunk 1]
```diff
@@ -113,14 +113,47 @@ void GetSkeletonBounds(const spSkeleton* skeleton, SpineModelBounds& bounds)
 
 }
 
+uint32_t CalcVertexBufferSize(const spSkeleton* skeleton, uint32_t* out_max_triangle_count)
```
Comment (JCash) [inline defold-spine/commonsrc/vertices.cpp:116]: Do one pass, calculating how much memory is needed

File: defold-spine/commonsrc/vertices.cpp [hunk 2]
```diff
@@ -113,14 +113,47 @@ void GetSkeletonBounds(const spSkeleton* skeleton, SpineModelBounds& bounds)
...
+    }
+    *out_max_triangle_count = max_triangle_count;
+    return count;
+}
+
 
 uint32_t GenerateVertexData(dmArray<SpineVertex>& vertex_buffer, const spSkeleton* skeleton, const dmVMath::Matrix4& world)
 {
-    dmArray<float> scratch; // scratch buffer
+    static dmArray<float> scratch; // scratch buffer
```
Comment (JCash) [inline defold-spine/commonsrc/vertices.cpp:147]: Not great, but I'll move this into an argument later.

File: defold-spine/commonsrc/vertices.cpp [hunk 3]
```diff
@@ -113,14 +113,47 @@ void GetSkeletonBounds(const spSkeleton* skeleton, SpineModelBounds& bounds)
...
 
     int vindex = vertex_buffer.Size();
     int vindex_start = vindex;
 
+    uint32_t max_triangle_count = 0;
+    uint32_t estimated_vcount = CalcVertexBufferSize(skeleton, &max_triangle_count);
+    EnsureArrayFitsNumber(scratch, max_triangle_count);
+    EnsureArrayFitsNumber(vertex_buffer, estimated_vcount);
```
Comment (JCash) [inline defold-spine/commonsrc/vertices.cpp:155]: Do the full allocations upfront (if necessary)

File: defold-spine/commonsrc/vertices.cpp [hunk 4]
```diff
@@ -241,10 +264,9 @@ uint32_t GenerateVertexData(dmArray<SpineVertex>& vertex_buffer, const spSkeleto
 
                 addVertex(&vertex_buffer[vindex++], scratch[index], scratch[index + 1], uvs[index], uvs[index + 1], colorR, colorG, colorB, colorA);
             }
-
-            scratch.SetSize(0);
         }
     }
+    scratch.SetSize(0);
```
Comment (JCash) [inline defold-spine/commonsrc/vertices.cpp:269]: If it's not reset, then it will grow in capacity

File: defold-spine/src/comp_spine_model.cpp [hunk 5]
```diff
@@ -860,7 +860,7 @@ namespace dmSpine
         {
             case dmRender::RENDER_LIST_OPERATION_BEGIN:
             {
-                dmGraphics::SetVertexBufferData(world->m_VertexBuffer, 0, 0, dmGraphics::BUFFER_USAGE_STATIC_DRAW);
+                dmGraphics::SetVertexBufferData(world->m_VertexBuffer, 0, 0, dmGraphics::BUFFER_USAGE_DYNAMIC_DRAW);
```
Comment (JCash) [inline defold-spine/src/comp_spine_model.cpp:863]: This is a dynamic vertex buffer that we update each frame

**Review by britzl — APPROVED** [review]
Looks good

**Pull request merged**